feat: Add `sidebar_options` for configuring the sidebar

gadenbuie opened the pull request:
Currently, (AFAICT) it's not possible to configure the sidebar generated by quartodoc.

This PR adds support `quartodoc.sidebar_options` which is then added to the `id` and `contents` created by quartodoc.

```yaml
quartodoc:
  package: quartodoc.tests.example
  sidebar: _reference.yml
  sidebar_options:
    style: docked
    search: true
```

This approach keeps sidebar options orthogonal to whatever choices about `sidebar` that quartodoc makes. It also ensures that `id` and `contents` are wholly controlled by quartodoc.

## Another option

Depending on how you feel about `sidebar_options`, I could certainly see a world where all of this is consolidated in `sidebar`, where we'd take all of the quarto options, plus an additional field like `file` for the yaml file name. Backwards compat would be possible by taking the string value

```yaml
quartodoc:
  sidebar: "_my-pkg.yml"
```

and automatically expanding to a full dictionary equivalent to

```yaml
quartodoc:
  sidebar:
    file: "_my-pkg.yml"
```

I think conceptually this is a bit nicer, but there's a small risk that we choose a field name for the sidebar filename that Quarto later chooses to add to `website.sidebar`. In this world, the first example would become something like this:

```yaml
quartodoc:
  package: quartodoc.tests.example
  sidebar: 
    file: _reference.yml
    style: docked
    search: true
```

If done this way, we could also let users choose `id` and customize `contents`. `id` is easily overwritten, but for `contents` we'd have to have a way to signal where the quartodoc contents should go.

My proposal for `contents` in this scenario would be to use a placeholder value to signal where to put quartodoc's contents, following these rules:

1. If no `contents`, use quartodoc's contents.
2. If `contents` and no sentinel value, append quartodoc's contents.
3. Place quartodoc's contents in a content item that matches `{{ contents }}`.

If you like this proposal, I don't think it would take long to rework this PR.

machow commented:
Thanks for taking the time to work on this! I'm onboard with either approach, so whatever you think is better! It seems like an advantage of having a single `sidebar:` config is we could address it on a single page easily (somehow it seems simpler than explaining there are two sidebar fields).

(RE quarto using a `file:`, seems okay to workaround if we ever come to it)

machow commented:
Here are some relevant issues I found:

* https://github.com/machow/quartodoc/discussions/330
  - https://github.com/quarto-dev/quarto-cli/issues/8449

gadenbuie commented:
@machow I've reworked the PR to use a dictionary for `quartodoc.sidebar`. I also rewrote the PR description above to give a better overview of the current state of the feature.

One place I could use your expertise in the review is making sure there aren't unwanted upstream consequences. I think I've caught everywhere that `self.sidebar` (previously a string or None) has become `self.sidebar["file"]`, but it could use another look.

gadenbuie commented on quartodoc/autosummary.py:
```diff
@@ -588,7 +598,7 @@ def build(self, filter: str = "*"):
         # sidebar ----
 
         if self.sidebar:
-            _log.info(f"Writing sidebar yaml to {self.sidebar}")
+            _log.info(f"Writing sidebar yaml to {self.sidebar['file']}")
             self.write_sidebar(blueprint)
```
Actually, I think I've convinced myself that everything is all good. Here's where we decide whether or not to write the sidebar, by which point `self.sidebar` is guaranteed to be a dictionary containing at least a `file` item (in `Builder.__init__` just above in the diff).

machow commented on quartodoc/autosummary.py:
```diff
@@ -686,14 +698,48 @@ def _generate_sidebar(self, blueprint: layout.Layout):
+            if not isinstance(sidebar["contents"], list):
+                raise TypeError("`sidebar.contents` must be a list")
+
+            def splice_contents_recursive(sidebar, contents):
```
Do you mind pulling this into a staticmethod on the class (or a function outside the class)? It feels like if it were called `_insert_contents(x: dict | list, contents: ?)` or similar, the idea of a utility function putting contents where a sentinel string is is a bit easier to think about outside of the context of the sidebar.

machow reviewed (changes requested):
Thanks, this looks really great! I left a small nit asking whether we could pull out the splice function onto the class or as a top-level function. It took a second to get the gist of what was happening, but as a general function it seems like it'd be quick to pick up on what's happening!

machow reviewed (approved):
Thanks for making the changes -- hope it's okay, I moved some of the advanced sidebar piece into its own section on the sidebar page

machow reviewed (approved):
Thanks for making the changes -- hope it's okay, I moved some of the advanced sidebar piece into its own section on the sidebar page
